File: android/app/src/main/java/com/neotun/android/config/UriParser.kt
package com.neotun.android.config

import com.neotun.android.models.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.URI
import java.net.URLDecoder
import java.util.Base64

class UriParser {
    
    fun parseUri(uri: String): VpnProfile? {
        return try {
            when {
                uri.startsWith("vmess://") -> parseVMessUri(uri)
                uri.startsWith("vless://") -> parseVLessUri(uri)
                uri.startsWith("trojan://") -> parseTrojanUri(uri)
                uri.startsWith("ss://") -> parseShadowsocksUri(uri)
                else -> null
            }
        } catch (e: Exception) {
            null
        }
    }
    
    private fun parseVMessUri(uri: String): VpnProfile? {
        val base64Data = uri.removePrefix("vmess://")
        val jsonString = String(Base64.getDecoder().decode(base64Data))
        
        // Parse VMess JSON format
        val json = Json { ignoreUnknownKeys = true }
        val vmessData = json.decodeFromString<VMessUriData>(jsonString)
        
        return VpnProfile(
            name = vmessData.ps ?: "VMess Server",
            protocol = VpnProtocol.VMESS,
            server = vmessData.add,
            port = vmessData.port.toInt(),
            credentials = VpnCredentials.VMess(
                userId = vmessData.id,
                alterId = vmessData.aid,
                security = vmessData.scy ?: "auto"
            ),
            settings = VpnSettings(
                network = vmessData.net ?: "tcp",
                security = vmessData.tls ?: "none",
                tlsSettings = if (vmessData.tls == "tls") {
                    TlsSettings(serverName = vmessData.sni)
                } else null,
                wsSettings = if (vmessData.net == "ws") {
                    WebSocketSettings(
                        path = vmessData.path ?: "/",
                        headers = vmessData.host?.let { mapOf("Host" to it) } ?: emptyMap()
                    )
                } else null
            )
        )
    }
    
    private fun parseVLessUri(uri: String): VpnProfile? {
        val parsedUri = URI(uri)
        val userId = parsedUri.userInfo
        val server = parsedUri.host
        val port = parsedUri.port
        
        val params = parseQueryParams(parsedUri.query ?: "")
        val fragment = URLDecoder.decode(parsedUri.fragment ?: "VLess Server", "UTF-8")
        
        return VpnProfile(
            name = fragment,
            protocol = VpnProtocol.VLESS,
            server = server,
            port = port,
            credentials = VpnCredentials.VLess(
                userId = userId,
                flow = params["flow"],
                encryption = params["encryption"] ?: "none"
            ),
            settings = VpnSettings(
                network = params["type"] ?: "tcp",
                security = params["security"] ?: "none",
                tlsSettings = if (params["security"] == "tls") {
                    TlsSettings(serverName = params["sni"])
                } else null,
                realitySettings = if (params["security"] == "reality") {
                    RealitySettings(
                        publicKey = params["pbk"] ?: "",
                        shortId = params["sid"] ?: "",
                        serverName = params["sni"] ?: "",
                        fingerprint = params["fp"] ?: "chrome"
                    )
                } else null
            )
        )
    }
    
    private fun parseTrojanUri(uri: String): VpnProfile? {
        val parsedUri = URI(uri)
        val password = parsedUri.userInfo
        val server = parsedUri.host
        val port = parsedUri.port
        
        val params = parseQueryParams(parsedUri.query ?: "")
        val fragment = URLDecoder.decode(parsedUri.fragment ?: "Trojan Server", "UTF-8")
        
        return VpnProfile(
            name = fragment,
            protocol = VpnProtocol.TROJAN,
            server = server,
            port = port,
            credentials = VpnCredentials.Trojan(password = password),
            settings = VpnSettings(
                network = params["type"] ?: "tcp",
                security = params["security"] ?: "tls",
                tlsSettings = TlsSettings(
                    serverName = params["sni"] ?: server,
                    allowInsecure = params["allowInsecure"] == "1"
                )
            )
        )
    }
    
    private fun parseShadowsocksUri(uri: String): VpnProfile? {
        val parsedUri = URI(uri)
        val userInfo = String(Base64.getDecoder().decode(parsedUri.userInfo))
        val parts = userInfo.split(":")
        
        if (parts.size != 2) return null
        
        val method = parts[0]
        val password = parts[1]
        val server = parsedUri.host
        val port = parsedUri.port
        val fragment = URLDecoder.decode(parsedUri.fragment ?: "Shadowsocks Server", "UTF-8")
        
        return VpnProfile(
            name = fragment,
            protocol = VpnProtocol.SHADOWSOCKS,
            server = server,
            port = port,
            credentials = VpnCredentials.Shadowsocks(
                method = method,
                password = password
            )
        )
    }
    
    private fun parseQueryParams(query: String): Map<String, String> {
        return query.split("&")
            .mapNotNull { param ->
                val parts = param.split("=", limit = 2)
                if (parts.size == 2) {
                    URLDecoder.decode(parts[0], "UTF-8") to URLDecoder.decode(parts[1], "UTF-8")
                } else null
            }
            .toMap()
    }
}

@Serializable
private data class VMessUriData(
    val v: String? = null,
    val ps: String? = null,
    val add: String,
    val port: String,
    val id: String,
    val aid: Int = 0,
    val scy: String? = null,
    val net: String? = null,
    val type: String? = null,
    val host: String? = null,
    val path: String? = null,
    val tls: String? = null,
    val sni: String? = null
)

File: windows/NeoTUN.Core/Config/UriParser.cs
using System.Text;
using System.Text.Json;
using NeoTUN.Core.Models;

namespace NeoTUN.Core.Config;

public class UriParser
{
    public VpnProfile? ParseUri(string uri)
    {
        try
        {
            return uri switch
            {
                var u when u.StartsWith("vmess://") => ParseVMessUri(u),
                var u when u.StartsWith("vless://") => ParseVLessUri(u),
                var u when u.StartsWith("trojan://") => ParseTrojanUri(u),
                var u when u.StartsWith("ss://") => ParseShadowsocksUri(u),
                _ => null
            };
        }
        catch
        {
            return null;
        }
    }

    private VpnProfile? ParseVMessUri(string uri)
    {
        var base64Data = uri["vmess://".Length..];
        var jsonBytes = Convert.FromBase64String(base64Data);
        var jsonString = Encoding.UTF8.GetString(jsonBytes);

        var vmessData = JsonSerializer.Deserialize<VMessUriData>(jsonString);
        if (vmessData == null) return null;

        return new VpnProfile(
            Guid.NewGuid().ToString(),
            vmessData.Ps ?? "VMess Server",
            VpnProtocol.VMess,
            vmessData.Add,
            int.Parse(vmessData.Port),
            new VpnCredentials.VMess(
                vmessData.Id,
                vmessData.Aid,
                vmessData.Scy ?? "auto"
            ),
            new VpnSettings(
                vmessData.Net ?? "tcp",
                vmessData.Tls ?? "none",
                vmessData.Tls == "tls" ? new TlsSettings(vmessData.Sni) : null,
                vmessData.Net == "ws" ? new WebSocketSettings(
                    vmessData.Path ?? "/",
                    string.IsNullOrEmpty(vmessData.Host) ? null : new Dictionary<string, string> { ["Host"] = vmessData.Host }
                ) : null
            ),
            DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        );
    }

    private VpnProfile? ParseVLessUri(string uri)
    {
        var parsedUri = new Uri(uri);
        var userId = parsedUri.UserInfo;
        var server = parsedUri.Host;
        var port = parsedUri.Port;

        var queryParams = ParseQueryParams(parsedUri.Query);
        var fragment = Uri.UnescapeDataString(parsedUri.Fragment?.TrimStart('#') ?? "VLess Server");

        return new VpnProfile(
            Guid.NewGuid().ToString(),
            fragment,
            VpnProtocol.VLess,
            server,
            port,
            new VpnCredentials.VLess(
                userId,
                queryParams.GetValueOrDefault("flow"),
                queryParams.GetValueOrDefault("encryption", "none")
            ),
            new VpnSettings(
                queryParams.GetValueOrDefault("type", "tcp"),
                queryParams.GetValueOrDefault("security", "none"),
                queryParams.GetValueOrDefault("security") == "tls" ? new TlsSettings(queryParams.GetValueOrDefault("sni")) : null,
                null,
                queryParams.GetValueOrDefault("security") == "reality" ? new RealitySettings(
                    queryParams.GetValueOrDefault("pbk", ""),
                    queryParams.GetValueOrDefault("sid", ""),
                    queryParams.GetValueOrDefault("sni", ""),
                    queryParams.GetValueOrDefault("fp", "chrome")
                ) : null
            ),
            DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        );
    }

    private VpnProfile? ParseTrojanUri(string uri)
    {
        var parsedUri = new Uri(uri);
        var password = parsedUri.UserInfo;
        var server = parsedUri.Host;
        var port = parsedUri.Port;

        var queryParams = ParseQueryParams(parsedUri.Query);
        var fragment = Uri.UnescapeDataString(parsedUri.Fragment?.TrimStart('#') ?? "Trojan Server");

        return new VpnProfile(
            Guid.NewGuid().ToString(),
            fragment,
            VpnProtocol.Trojan,
            server,
            port,
            new VpnCredentials.Trojan(password),
            new VpnSettings(
                queryParams.GetValueOrDefault("type", "tcp"),
                queryParams.GetValueOrDefault("security", "tls"),
                new TlsSettings(
                    queryParams.GetValueOrDefault("sni", server),
                    queryParams.GetValueOrDefault("allowInsecure") == "1"
                )
            ),
            DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        );
    }

    private VpnProfile? ParseShadowsocksUri(string uri)
    {
        var parsedUri = new Uri(uri);
        var userInfoBytes = Convert.FromBase64String(parsedUri.UserInfo);
        var userInfo = Encoding.UTF8.GetString(userInfoBytes);
        var parts = userInfo.Split(':', 2);

        if (parts.Length != 2) return null;

        var method = parts[0];
        var password = parts[1];
        var server = parsedUri.Host;
        var port = parsedUri.Port;
        var fragment = Uri.UnescapeDataString(parsedUri.Fragment?.TrimStart('#') ?? "Shadowsocks Server");

        return new VpnProfile(
            Guid.NewGuid().ToString(),
            fragment,
            VpnProtocol.Shadowsocks,
            server,
            port,
            new VpnCredentials.Shadowsocks(method, password),
            new VpnSettings(),
            DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        );
    }

    private Dictionary<string, string> ParseQueryParams(string query)
    {
        var result = new Dictionary<string, string>();
        
        if (string.IsNullOrEmpty(query)) return result;
        
        query = query.TrimStart('?');
        
        foreach (var param in query.Split('&'))
        {
            var parts = param.Split('=', 2);
            if (parts.Length == 2)
            {
                var key = Uri.UnescapeDataString(parts[0]);
                var value = Uri.UnescapeDataString(parts[1]);
                result[key] = value;
            }
        }

        return result;
    }

    private class VMessUriData
    {
        public string? V { get; set; }
        public string? Ps { get; set; }
        public string Add { get; set; } = "";
        public string Port { get; set; } = "";
        public string Id { get; set; } = "";
        public int Aid { get; set; }
        public string? Scy { get; set; }
        public string? Net { get; set; }
        public string? Type { get; set; }
        public string? Host { get; set; }
        public string? Path { get; set; }
        public string? Tls { get; set; }
        public string? Sni { get; set; }
    }
}